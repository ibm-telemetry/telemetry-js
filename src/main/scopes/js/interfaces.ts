/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { JsNodeHandler } from './node-handlers/js-node-handler.js'
import { NodeValueHandler } from './node-handlers/value-handlers/node-value-handler.js'

export interface JsImport {
  name: string
  path: string
  isDefault: boolean
  isAll: boolean
  rename?: string
}

type JsNodeHandlerClass<DataType> = new (
  node: ts.SourceFile,
  logger: Logger
) => JsNodeHandler<DataType>

export type JsNodeHandlerMap = Partial<Record<ts.SyntaxKind, JsNodeHandlerClass<unknown>>>

export type NodeValue = string | number | boolean | ComplexValue | null | undefined

type NodeValueHandlerProducer = new (node: ts.SourceFile, logger: Logger) => NodeValueHandler

export type NodeValueHandlerMap = Partial<Record<ts.SyntaxKind, NodeValueHandlerProducer>>
