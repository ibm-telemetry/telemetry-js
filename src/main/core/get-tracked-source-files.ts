/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { readFile } from 'node:fs/promises'
import * as path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from './log/logger.js'
import { TrackedFileEnumerator } from './tracked-file-enumerator.js'
import { parseDocument } from 'htmlparser2'
import type { Document as HtmlDocument } from 'domhandler'
import { HtmlParsedFile } from '../scopes/wc/interfaces.js'

/**
 * Gets all tracked source files to consider for data collection,
 * filtered by supplied file extension array.
 *
 * @param cwd - Working directory for underlying command execution.
 * @param root - Root directory in which to search for tracked source files. This is an absolute
 * path.
 * @param logger - Logger instance to use.
 * @param fileExtensions - List of file extensions to filter files by.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(
  cwd: string,
  root: string,
  logger: Logger,
  fileExtensions: string[]
) {
  logger.traceEnter('', 'getTrackedSourceFiles', [root, fileExtensions])

  const fileEnumerator = new TrackedFileEnumerator(logger)
  const files = []

  // If a file is passed instead of a directory, avoid the `git ls-tree` call
  if (fileExtensions.includes(path.extname(root))) {
    files.push(root)
  } else {
    files.push(
      ...(await fileEnumerator.find(cwd, root, (file) =>
        fileExtensions.includes(path.extname(file))
      ))
    )
  }

  const promises = files.map(async (file) => {
    return {
      fileName: file,
      async createSourceFile(): Promise<ts.SourceFile | HtmlDocument> {
        // Parse html file with `htmlparser2
        if (file.endsWith('html') || file.endsWith('htm')) {
          const parsedFile = parseDocument((await readFile(file)).toString()) as HtmlParsedFile
          parsedFile.fileName = file
        }

        // Parse any other files
        return ts.createSourceFile(
          file,
          (await readFile(file)).toString(),
          ts.ScriptTarget.ES2021,
          /* setParentNodes */ true
        )
      }
    }
  })

  const results = await Promise.all(promises)

  logger.traceExit(
    '',
    'getTrackedSourceFiles',
    results.map((result) => result.fileName)
  )
  return results
}
