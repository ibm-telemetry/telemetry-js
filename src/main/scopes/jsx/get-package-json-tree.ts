/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import fs from 'fs/promises'
import path from 'path'

import { DirectoryEnumerator } from '../../core/directory-enumerator.js'
import { type Logger } from '../../core/log/logger.js'
import { type FileTree } from './interfaces.js'

async function hasPackageJsonFile(dir: string) {
  const packageJsonPath = path.join(dir, 'package.json')

  try {
    await fs.access(packageJsonPath)
  } catch {
    return false
  }

  return true
}

/**
 * Constructs an array of trees from all package.json files in the given repository,
 *  where every nested object is a subpath of the root.
 *
 * @param root - Root-most directory to consider.
 * @param cwd - Current working directory to find tracked files for.
 * @param logger - Logger instance.
 * @returns Array of trees of package.json files.
 */
export async function getPackageJsonTree(
  root: string,
  cwd: string,
  logger: Logger
): Promise<FileTree[]> {
  const dirs = (await new DirectoryEnumerator(logger).find(cwd, root, hasPackageJsonFile)).sort(
    (a, b) => a.split('/').length - b.split('/').length
  )

  const tree: FileTree[] = []

  dirs.forEach((path) => {
    let currNode = tree
    let nextNode = tree.find((root) => path.includes(root.path))
    while (nextNode !== null && nextNode !== undefined) {
      currNode = nextNode.children
      nextNode = currNode.find((node) => path.includes(node.path))
    }
    currNode.push({ path, children: [] })
  })

  return tree
}
