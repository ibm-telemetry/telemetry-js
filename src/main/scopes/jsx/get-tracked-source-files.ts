/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from '../../core/log/logger.js'
import { TrackedFileEnumerator } from '../../core/tracked-file-enumerator.js'

/**
 * Gets all tracked source files to consider for data collection.
 *
 * @param root - Root directory in which to search for tracked source files.
 * @param logger - Logger instance to use.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(root: string, logger: Logger) {
  const fileEnumerator = new TrackedFileEnumerator(logger)
  const allowedSuffixes = ['js', 'jsx', 'ts', 'tsx']

  const files = await fileEnumerator.find(root, (file) =>
    allowedSuffixes.includes(path.extname(file))
  )

  const program = ts.createProgram(files, {})

  return program.getSourceFiles().filter((file) => !file.isDeclarationFile)
}
