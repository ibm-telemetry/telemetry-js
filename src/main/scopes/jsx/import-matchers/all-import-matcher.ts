/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { type JsxElement, type JsxElementImportMatcher, type JsxImport } from '../interfaces.js'

/**
 * Identifies JsxElements that have been imported as all imports,
 * and returns an import element match (if any) or undefined otherwise.
 */
export class AllImportMatcher implements JsxElementImportMatcher {
  /**
   * Determines if a given JsxElement is an all(*) import
   * (.e.g: import * as something from 'package')
   * that matches the supplied list of import elements.
   *
   * @param element - JsxElement to evaluate.
   * @param imports - Import elements to use for comparison.
   * @returns Corresponding JsxImportElement if element was imported as an all import,
   * undefined otherwise.
   */
  findMatch(element: JsxElement, imports: JsxImport[]) {
    return element.prefix !== undefined
      ? imports.find((i) => i.isAll && i.name === element.prefix)
      : undefined
  }
}
