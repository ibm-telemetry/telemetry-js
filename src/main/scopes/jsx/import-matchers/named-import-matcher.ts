/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { type JsxElement, type JsxElementImportMatcher, type JsxImport } from '../interfaces.js'

/**
 * Identifies JsxElements that have been imported as named imports,
 * and returns an import element match (if any) or undefined otherwise.
 */
export class NamedImportMatcher implements JsxElementImportMatcher {
  /**
   * Determines if a given JsxElement is a named import (e.g.: `import {something} from 'package'`).
   *
   * @param element - JsxElement to evaluate.
   * @param imports - Import elements to use for comparison.
   * @returns Corresponding JsxImport if element was imported as a name import,
   * undefined otherwise.
   */
  findMatch(element: JsxElement, imports: JsxImport[]) {
    return element.prefix === undefined
      ? imports.find((i) => !i.isDefault && !i.isAll && i.name === element.name)
      : undefined
  }
}
