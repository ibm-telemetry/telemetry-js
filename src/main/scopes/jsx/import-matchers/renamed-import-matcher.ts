/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  type JsxElementImportMatcher,
  type JsxImportElement,
  type PartialJsxElement
} from '../interfaces.js'

/**
 * Identifies JsxElements that have been imported as renamed imports.
 */
export class RenamedImportMatcher implements JsxElementImportMatcher {
  /**
   * Determines if a given JsxElement is a renamed import
   * (e.g.: import {something as somethingElse} from 'package).
   *
   * @param element - JsxElement to evaluate.
   * @param imports - Import elements to use for comparison.
   * @returns True if element was imported a named import, false otherwise.
   */
  isMatch(element: PartialJsxElement, imports: JsxImportElement[]) {
    return element.prefix !== undefined && imports.some((i) => i.rename === element.prefix)
  }

  /**
   * Sanitizes the contents of a given JsxElement.
   *
   * @param element - JsxElement to sanitize.
   * @returns Sanitized version of element (in this case, the element itself).
   */
  getSanitizedElement(element: PartialJsxElement) {
    return element
  }
}
