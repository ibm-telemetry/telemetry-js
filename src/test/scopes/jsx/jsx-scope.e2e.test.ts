/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'node:path'

import { type ConfigSchema } from '@ibm/telemetry-config-schema'
import type * as ts from 'typescript'
import { describe, expect, it } from 'vitest'

import { EmptyScopeError } from '../../../main/exceptions/empty-scope.error.js'
import { AllImportMatcher } from '../../../main/scopes/jsx/import-matchers/all-import-matcher.js'
import { NamedImportMatcher } from '../../../main/scopes/jsx/import-matchers/named-import-matcher.js'
import { RenamedImportMatcher } from '../../../main/scopes/jsx/import-matchers/renamed-import-matcher.js'
import { FileTree } from '../../../main/scopes/jsx/interfaces.js'
import { JsxElementAccumulator } from '../../../main/scopes/jsx/jsx-element-accumulator.js'
import { JsxScope } from '../../../main/scopes/jsx/jsx-scope.js'
import { getPackageJsonTree } from '../../../main/scopes/jsx/utils/get-package-json-tree.js'
import { getTrackedSourceFiles } from '../../../main/scopes/jsx/utils/get-tracked-source-files.js'
import { clearDataPointTimes } from '../../__utils/clear-data-point-times.js'
import { clearTelemetrySdkVersion } from '../../__utils/clear-telemetry-sdk-version.js'
import { Fixture } from '../../__utils/fixture.js'
import { initLogger } from '../../__utils/init-logger.js'
import { initializeOtelForTest } from '../../__utils/initialize-otel-for-test.js'

const config: ConfigSchema = {
  projectId: 'abc123',
  version: 1,
  endpoint: '',
  collect: {
    jsx: {
      elements: {
        allowedAttributeNames: ['firstProp', 'secondProp'],
        allowedAttributeStringValues: ['hi', 'wow']
      }
    }
  }
}

describe('class: JsxScope', () => {
  const logger = initLogger()
  describe('run', () => {
    it('correctly captures jsx element metric data', async () => {
      const metricReader = initializeOtelForTest()
      const root = new Fixture(path.join('projects', 'basic-project'))
      const cwd = new Fixture(
        path.join('projects', 'basic-project', 'node_modules', 'instrumented')
      )
      const jsxScope = new JsxScope(cwd.path, root.path, config, logger)

      jsxScope.setRunSync()
      await jsxScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })

    it("does not capture metrics for files that are not in instrumented package's installer", async () => {
      const metricReader = initializeOtelForTest()
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const cwd = new Fixture(
        path.join('projects', 'complex-nesting-thingy', 'node_modules', 'instrumented')
      )
      const jsxScope = new JsxScope(cwd.path, root.path, config, logger)

      jsxScope.setRunSync()
      await jsxScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })

    it('captures metrics when instrumented package is installed in intermediate package', async () => {
      const metricReader = initializeOtelForTest()
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const cwd = new Fixture(
        path.join(
          'projects',
          'complex-nesting-thingy',
          'node_modules',
          'intermediate',
          'node_modules',
          'instrumented'
        )
      )
      const jsxScope = new JsxScope(cwd.path, root.path, config, logger)

      jsxScope.setRunSync()
      await jsxScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })

    it('captures metrics for nested files when instrumented package is installed in top-level package', async () => {
      const metricReader = initializeOtelForTest()
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const cwd = new Fixture(
        path.join('projects', 'complex-nesting-thingy', 'node_modules', 'instrumented-top-level')
      )
      const jsxScope = new JsxScope(cwd.path, root.path, config, logger)

      jsxScope.setRunSync()
      await jsxScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })

    it('does not capture metrics for nested files when local package has installed a more specific version of the instrumented package', async () => {
      const metricReader = initializeOtelForTest()
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const cwd = new Fixture(
        path.join('projects', 'complex-nesting-thingy', 'node_modules', 'another-package')
      )
      const jsxScope = new JsxScope(cwd.path, root.path, config, logger)

      jsxScope.setRunSync()
      await jsxScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })

    it('throws EmptyScopeError if no collector has been defined', async () => {
      const fixture = new Fixture(
        path.join('projects', 'basic-project', 'node_modules', 'instrumented')
      )
      const jsxScope = new JsxScope(
        fixture.path,
        path.join(fixture.path, '..', '..'),
        { collect: { npm: {} }, projectId: '123', version: 1, endpoint: '' },
        logger
      )

      jsxScope.setRunSync()
      await expect(jsxScope.run()).rejects.toThrow(EmptyScopeError)
    })
  })

  describe('processFile', () => {
    const fixture = new Fixture(path.join('projects', 'basic-project', 'test.jsx'))
    const root = new Fixture(path.join('projects', 'basic-project'))
    const jsxScope = new JsxScope(fixture.path, root.path, config, logger)

    it('correctly detects imports and elements in a given file', async () => {
      const accumulator = new JsxElementAccumulator()
      const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile

      jsxScope.processFile(accumulator, sourceFile)
      expect(accumulator.elements).toMatchSnapshot('elements')
      expect(accumulator.imports).toMatchSnapshot('imports')
    })
  })

  describe('findFileLocalInstaller', () => {
    it('correctly finds direct installer for file', async () => {
      const fixture = new Fixture(path.join('projects', 'basic-project', 'test.jsx'))
      const root = new Fixture(path.join('projects', 'basic-project'))
      const jsxScope = new JsxScope(fixture.path, root.path, config, logger)
      const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile

      const localInstaller = { name: 'basic-project', version: '1.0.0' }

      const fileLocalInstaller = await jsxScope.findFileLocalInstaller(sourceFile, 'instrumented', [
        localInstaller
      ])
      expect(fileLocalInstaller).toStrictEqual(localInstaller)
    })

    it('correctly finds top-level installer for file', async () => {
      const fixture = new Fixture(
        path.join('projects', 'complex-nesting-thingy', 'package1', 'test.jsx')
      )
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const jsxScope = new JsxScope(fixture.path, root.path, config, logger)
      const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile

      const localInstaller = { name: 'complex-nesting-thingy', version: '1.0.0' }

      const fileLocalInstaller = await jsxScope.findFileLocalInstaller(
        sourceFile,
        'instrumented-top-level',
        [localInstaller]
      )
      expect(fileLocalInstaller).toStrictEqual(localInstaller)
    })

    it('correctly returns undefined for file with no local installer', async () => {
      const fixture = new Fixture(path.join('projects', 'basic-project', 'test.jsx'))
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const jsxScope = new JsxScope(fixture.path, root.path, config, logger)
      const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile

      const localInstaller = { name: 'complex-nesting-thingy', version: '1.0.0' }

      const fileLocalInstaller = await jsxScope.findFileLocalInstaller(
        sourceFile,
        'instrumented-top-level',
        [localInstaller]
      )
      expect(fileLocalInstaller).toBeUndefined()
    })

    it('correctly returns undefined for file that has a different immediate local install', async () => {
      const fixture = new Fixture(
        path.join('projects', 'complex-nesting-thingy', 'package1', 'test.jsx')
      )
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const jsxScope = new JsxScope(fixture.path, root.path, config, logger)
      const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile

      const localInstaller = { name: 'complex-nesting-thingy', version: '1.0.0' }

      const fileLocalInstaller = await jsxScope.findFileLocalInstaller(
        sourceFile,
        'another-package',
        [localInstaller]
      )
      expect(fileLocalInstaller).toBeUndefined()
    })

    it('correctly returns undefined for file that has a mid-level different local install', async () => {
      const fixture = new Fixture(
        path.join('projects', 'complex-nesting-thingy', 'mid-level', 'nested-level', 'test.jsx')
      )
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const jsxScope = new JsxScope(fixture.path, root.path, config, logger)
      const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile

      const localInstaller = { name: 'complex-nesting-thingy', version: '1.0.0' }

      const fileLocalInstaller = await jsxScope.findFileLocalInstaller(
        sourceFile,
        'another-package',
        [localInstaller]
      )
      expect(fileLocalInstaller).toBeUndefined()
    })
  })

  describe('removeIrrelevantImports', () => {
    it('correctly removes unwanted imports', () => {
      const accumulator = new JsxElementAccumulator()
      accumulator.imports.push({
        name: 'name',
        path: 'path',
        isDefault: false,
        isAll: false
      })
      accumulator.imports.push({
        name: 'name',
        path: 'instrumented',
        isDefault: true,
        isAll: false
      })
      accumulator.imports.push({
        name: 'name',
        path: 'instrumented/bla',
        isDefault: false,
        isAll: true
      })
      accumulator.imports.push({
        name: 'name',
        path: 'instrumented-bla',
        isDefault: false,
        isAll: true
      })
      expect(accumulator.imports).toHaveLength(4)
      const jsxScope = new JsxScope('', '', config, logger)
      jsxScope.removeIrrelevantImports(accumulator, 'instrumented')
      expect(accumulator.imports).toHaveLength(2)
    })

    it('removes all imports if no imports match instrumented package', () => {
      const accumulator = new JsxElementAccumulator()
      accumulator.imports.push({
        name: 'name',
        path: 'path',
        isDefault: false,
        isAll: false
      })
      accumulator.imports.push({
        name: 'name',
        path: 'not-instrumented',
        isDefault: true,
        isAll: false
      })
      accumulator.imports.push({
        name: 'name',
        path: 'not-instrumented/bla',
        isDefault: false,
        isAll: true
      })
      expect(accumulator.imports).toHaveLength(3)
      const jsxScope = new JsxScope('', '', config, logger)
      jsxScope.removeIrrelevantImports(accumulator, 'instrumented')
      expect(accumulator.imports).toHaveLength(0)
    })

    it("does not remove imports if there aren't any to remove", () => {
      const accumulator = new JsxElementAccumulator()
      accumulator.imports.push({
        name: 'name',
        path: 'instrumented/1',
        isDefault: false,
        isAll: false
      })
      accumulator.imports.push({
        name: 'name',
        path: 'instrumented',
        isDefault: true,
        isAll: false
      })
      accumulator.imports.push({
        name: 'name',
        path: 'instrumented/bla',
        isDefault: false,
        isAll: true
      })
      expect(accumulator.imports).toHaveLength(3)
      const jsxScope = new JsxScope('', '', config, logger)
      jsxScope.removeIrrelevantImports(accumulator, 'instrumented')
      expect(accumulator.imports).toHaveLength(3)
    })

    it('can accept empty array', () => {
      const accumulator = new JsxElementAccumulator()
      expect(accumulator.imports).toHaveLength(0)
      const jsxScope = new JsxScope('', '', config, logger)
      jsxScope.removeIrrelevantImports(accumulator, 'instrumented')
      expect(accumulator.imports).toHaveLength(0)
    })
  })

  describe('resolveElementImports', () => {
    const jsxScope = new JsxScope('', '', config, logger)
    const defaultImport = {
      name: '[Default]',
      rename: 'nameDefault',
      path: 'instrumented',
      isDefault: true,
      isAll: false
    }
    const allImport = {
      name: 'all',
      path: 'instrumented',
      isDefault: false,
      isAll: true
    }
    const namedImport = {
      name: 'name',
      path: 'instrumented',
      isDefault: false,
      isAll: false
    }
    const renamedImport = {
      name: 'renameName',
      rename: 'rename',
      path: 'instrumented',
      isDefault: false,
      isAll: false
    }
    const defaultElement = {
      name: 'nameDefault',
      prefix: undefined,
      raw: '',
      attributes: []
    }
    const allElement = {
      name: 'whatever',
      prefix: 'all',
      raw: '',
      attributes: []
    }
    const namedElement = {
      name: 'name',
      prefix: undefined,
      raw: '',
      attributes: []
    }
    const renamedElement = {
      name: 'rename',
      prefix: undefined,
      raw: '',
      attributes: []
    }

    it('correctly identifies elements with their matchers', () => {
      const accumulator = new JsxElementAccumulator()
      accumulator.imports.push(defaultImport)
      accumulator.imports.push(allImport)
      accumulator.imports.push(namedImport)
      accumulator.imports.push(renamedImport)
      accumulator.elements.push(defaultElement)
      accumulator.elements.push(allElement)
      accumulator.elements.push(namedElement)
      accumulator.elements.push(renamedElement)

      jsxScope.resolveElementImports(accumulator, [
        new AllImportMatcher(),
        new NamedImportMatcher(),
        new RenamedImportMatcher()
      ])

      expect(accumulator.elementImports.get(defaultElement)).toStrictEqual(defaultImport)
      expect(accumulator.elementImports.get(allElement)).toStrictEqual(allImport)
      expect(accumulator.elementImports.get(namedElement)).toStrictEqual(namedImport)
      expect(accumulator.elementImports.get(renamedElement)).toStrictEqual(renamedImport)
    })

    it('discards elements that do not have a matcher', () => {
      const unmatchedElement1 = {
        name: 'noMatch1',
        prefix: undefined,
        raw: '',
        attributes: []
      }
      const unmatchedElement2 = {
        name: 'noMatch2',
        prefix: undefined,
        raw: '',
        attributes: []
      }
      const accumulator = new JsxElementAccumulator()
      accumulator.imports.push(defaultImport)
      accumulator.imports.push(allImport)
      accumulator.imports.push(namedImport)
      accumulator.imports.push(renamedImport)
      accumulator.elements.push(defaultElement)
      accumulator.elements.push(namedElement)
      accumulator.elements.push(renamedElement)
      accumulator.elements.push(unmatchedElement1)
      accumulator.elements.push(unmatchedElement2)

      const jsxScope = new JsxScope('', '', config, logger)
      jsxScope.resolveElementImports(accumulator, [
        new AllImportMatcher(),
        new NamedImportMatcher(),
        new RenamedImportMatcher()
      ])

      expect(accumulator.elementImports.get(defaultElement)).toStrictEqual(defaultImport)
      expect(accumulator.elementImports.get(namedElement)).toStrictEqual(namedImport)
      expect(accumulator.elementImports.get(renamedElement)).toStrictEqual(renamedImport)
      expect(accumulator.elementImports.get(unmatchedElement1)).toBeUndefined()
      expect(accumulator.elementImports.get(unmatchedElement2)).toBeUndefined()
    })

    it('can accept empty array', () => {
      const accumulator = new JsxElementAccumulator()
      expect(() => {
        jsxScope.resolveElementImports(accumulator, [])
      }).not.toThrow()
    })
  })

  describe('resolveInvokers', () => {
    const jsxScope = new JsxScope('', '', config, logger)
    const element1 = {
      name: 'elName',
      prefix: undefined,
      raw: '',
      attributes: []
    }
    const element2 = {
      name: 'elName2',
      prefix: undefined,
      raw: '',
      attributes: []
    }

    it('correctly sets invoker name for elements', async () => {
      const accumulator = new JsxElementAccumulator()
      accumulator.elements.push(element1)
      accumulator.elements.push(element2)

      await jsxScope.resolveInvokers(accumulator, { name: 'basic-project', version: '0.0.1' })

      expect(accumulator.elementInvokers.get(element1)).toBe('basic-project')
      expect(accumulator.elementInvokers.get(element2)).toBe('basic-project')
    })
  })

  describe('findLocalPackages', () => {
    const jsxScope = new JsxScope('', '', config, logger)
    it('correctly finds local packages', async () => {
      const packagesRoot = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const tree = await getPackageJsonTree(packagesRoot.path, logger)

      await expect(jsxScope.findLocalPackages(tree)).resolves.toMatchSnapshot()
    })
    it('can tolerate packages that do not exist', async () => {
      const trees: FileTree[] = [
        {
          path: 'does/not/exist',
          children: []
        }
      ]

      await expect(jsxScope.findLocalPackages(trees)).resolves.toStrictEqual([])
    })
  })

  describe('findPkgLocalInstallers', () => {
    it('correctly finds installers for a top-level package', async () => {
      const root = new Fixture(path.join('projects', 'basic-project'))
      const cwd = new Fixture(path.join('projects', 'basic-project'))
      const jsxScope = new JsxScope(cwd.path, root.path, config, logger)
      const localPackages = [{ name: 'basic-project', version: '1.0.0' }]

      const installers = await jsxScope.findPkgLocalInstallers('foo', '1.0.0', localPackages)

      expect(installers).toMatchSnapshot()
    })
    it('correctly finds installers for a given package that was installed in an intermediate package', async () => {
      const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
      const cwd = new Fixture(
        path.join(
          'projects',
          'complex-nesting-thingy',
          'node_modules',
          'intermediate',
          'node_modules',
          'instrumented'
        )
      )
      const jsxScope = new JsxScope(cwd.path, root.path, config, logger)
      const localPackages = [
        { name: 'package1', version: '1.0.0' },
        { name: 'package2', version: '1.0.0' },
        { name: 'package3', version: '1.0.0' },
        { name: 'complex-nesting-thingy', version: '1.0.0' }
      ]

      const installers = await jsxScope.findPkgLocalInstallers(
        'instrumented',
        '2.0.0',
        localPackages
      )

      expect(installers).toMatchSnapshot()
    }, 10000)
    it('does not find any installers for unknown package', async () => {
      const jsxScope = new JsxScope('', '', config, logger)
      const localPackages = [{ name: 'basic-project', version: '1.0.0' }]

      const installers = await jsxScope.findPkgLocalInstallers(
        'non-existent',
        '1.0.0',
        localPackages
      )

      expect(installers).toStrictEqual([])
    })
  })
})
