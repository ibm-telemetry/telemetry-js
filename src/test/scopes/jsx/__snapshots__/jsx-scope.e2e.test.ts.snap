// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: JsxScope > parseFile > correctly detects imports and elements in a given file 1`] = `
[
  {
    "attributes": [
      {
        "name": "something",
        "value": "1 === 5 ? 'boo' : 'baa'",
      },
      {
        "name": "simple",
        "value": "\\"4\\"",
      },
    ],
    "name": "Button",
    "prefix": undefined,
    "raw": "<Button something={1 === 5 ? 'boo' : 'baa'} simple=\\"4\\" hi={} woo>Hey</Button>",
  },
]
`;

exports[`class: JsxScope > parseFile > correctly detects imports and elements in a given file 2`] = `
[
  {
    "isAll": false,
    "isDefault": true,
    "name": "[Default]",
    "path": "instrumented",
    "rename": "Button",
  },
]
`;

exports[`class: JsxScope > run > correctly captures jsx element metric data 1`] = `
{
  "errors": [],
  "resourceMetrics": {
    "resource": Resource {
      "_asyncAttributesPromise": undefined,
      "_attributes": {
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.raw": "gitOrigin",
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "project.id": "projectId",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.17.1",
      },
      "_syncAttributes": {
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.raw": "gitOrigin",
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "project.id": "projectId",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "1.17.1",
      },
      "asyncAttributesPending": false,
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "attribute.names": [
                    "something",
                    "simple",
                  ],
                  "attribute.values": [
                    "1 === 5 ? 'boo' : 'baa'",
                    "\\"4\\"",
                  ],
                  "invoker.package.name": "basic-project",
                  "invoker.package.owner": undefined,
                  "invoker.package.raw": "basic-project",
                  "module.specifier": "instrumented",
                  "name": "[Default]",
                  "raw": "<Button something={1 === 5 ? 'boo' : 'baa'} simple=\\"4\\" hi={} woo>Hey</Button>",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 1,
              },
              {
                "attributes": {
                  "attribute.names": [
                    "firstProp",
                    "secondProp",
                    "numberProp",
                    "undefinedProp",
                    "nullProp",
                    "objectProp",
                    "isExplicitlyCool",
                    "isNotCool",
                    "objAccess",
                  ],
                  "attribute.values": [
                    "\\"hi\\"",
                    "\\"wow\\"",
                    "5",
                    "undefined",
                    "null",
                    "{key: 'value'}",
                    "true",
                    "false",
                    "obj['accessKey']",
                  ],
                  "invoker.package.name": "basic-project",
                  "invoker.package.owner": undefined,
                  "invoker.package.raw": "basic-project",
                  "module.specifier": "instrumented",
                  "name": "ImaginaryThing",
                  "raw": "<ImaginaryThing
          firstProp=\\"hi\\"
          secondProp=\\"wow\\"
          numberProp={5}
          undefinedProp={undefined}
          nullProp={null}
          objectProp={{key: 'value'}}
          isCool
          isExplicitlyCool={true}
          isNotCool={false}
          objAccess={obj['accessKey']}
          {...spreadObj}
        />",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 1,
              },
            ],
            "descriptor": {
              "description": "",
              "name": "element.count",
              "type": "COUNTER",
              "unit": "",
              "valueType": 0,
            },
            "isMonotonic": true,
          },
        ],
        "scope": {
          "name": "jsx",
          "schemaUrl": undefined,
          "version": "",
        },
      },
    ],
  },
}
`;
